crates/solver/src/driver.rs

```diff
@@ -145,13 +148,70 @@ impl Driver {
         .await
     }
 
+    async fn optimize_cost_of_unwrapping(&self, settlement: &mut Settlement) -> Result<()> {
+        let eth_for_settlement = settlement
+            .encoder
+            .unwraps()
+            .iter()
+            .find(|unwrap| unwrap.weth.address() == self.native_token)
+            .map(|unwrap| unwrap.amount)
+            .unwrap_or(U256([0; 4]));
+
+        let buffered_eth = self
+            .web3
+            .eth()
+            .balance(self.settlement_contract.address(), None)
+            .await?;
+
+        if buffered_eth >= eth_for_settlement {
+            // No need to unwrap yet, we have enough ETH buffered
+            tracing::debug!(?eth_for_settlement, "use internal buffer for unwraps");
+            settlement.encoder.drop_unwrap(self.native_token);
+            return Ok(());
+        }
+
+        let weth = contracts::WETH9::at(&self.web3, self.native_token);
+        let weth_balance = weth
+            .methods()
+            .balance_of(self.settlement_contract.address())
+            .call()
+            .await?;
+
+        let amount_to_unwrap =
+            U256::from_f64_lossy(weth_balance.to_f64_lossy() * self.weth_unwrap_factor);
+        if amount_to_unwrap > eth_for_settlement {
+            // If we unwrap more than this settlement requires, we might not need to unwrap for the next one
+            // which cuts the cost of 1 or more unwraps.
+            tracing::debug!(
+                ?amount_to_unwrap,
+                "unwrap parts of the settlement contract's WETH buffer"
+            );
+            settlement.encoder.drop_unwrap(self.native_token);
+            settlement
+                .encoder
+                .add_unwrap(crate::interactions::UnwrapWethInteraction {
+                    weth,
+                    amount: amount_to_unwrap,
+                });
+        } else {
+            tracing::warn!("consider topping up settlement contract's internal buffer");
```

Comment:
Not needed for now, but maybe a metric by how much we would have had to top up the contract could be useful to quantify the correct amount to use on average (can maybe be done using Dune as well)